File: sdk/derive/world.go
package derive

import "github.com/bpprotocol/blockparty/sdk/crypto"

// GlobalSalt is the fixed, non-secret domain-separation salt used to open a
// World from a seed phrase. It namespaces BlockParty derivations away from
// other systems that might reuse the same seed phrases.
const GlobalSalt = "bpprotocol.org/v1/global"

// World is a cryptographic domain: a ML-DSA signing key (the world_sig
// authority) plus the three salts that scope identity, type, and audience
// derivations.
type World struct {
	SigningKey   crypto.MLDSAKeyPair
	WalletSalt   []byte // scopes identity derivation
	TypeSalt     []byte // scopes type codes
	AudienceSalt []byte // scopes audience codes
}

// OpenWorld deterministically derives a World from a seed phrase. Anyone who
// knows the phrase derives the identical World, including its signing key.
func OpenWorld(seedPhrase string) World {
	worldSeed := crypto.HMACSHA256([]byte(GlobalSalt), []byte(seedPhrase))
	return GenerateWorld(crypto.MakeMLDSAPair(worldSeed), worldSeed)
}

// GenerateWorld builds a World from a signing key and the world seed. It is
// exposed to mirror the spec; most callers use OpenWorld.
func GenerateWorld(signingKey crypto.MLDSAKeyPair, worldSeed []byte) World {
	return World{
		SigningKey:   signingKey,
		WalletSalt:   crypto.HMACSHA256(worldSeed, []byte("wallets")),
		TypeSalt:     crypto.HMACSHA256(worldSeed, []byte("types")),
		AudienceSalt: crypto.HMACSHA256(worldSeed, []byte("channels")),
	}
}

File: implementations/typescript/packages/sdk/src/derive/world.ts
import { hmacSha256, makeMldsaPair, type KeyPair } from "../crypto/index.js";
import { utf8ToBytes } from "../bytes.js";

/**
 * Fixed, non-secret domain-separation salt used to open a World from a seed
 * phrase. Hard-coded into every conforming client.
 */
export const GLOBAL_SALT = "bpprotocol.org/v1/global";

/**
 * A cryptographic domain: an ML-DSA-65 signing key (the world_sig authority)
 * plus the three salts that scope identity, type, and audience derivations.
 */
export interface World {
  readonly signingKey: KeyPair;
  readonly walletSalt: Uint8Array;
  readonly typeSalt: Uint8Array;
  readonly audienceSalt: Uint8Array;
}

/**
 * Deterministically derive a World from a seed phrase. Anyone who knows the
 * phrase derives the identical World, including its signing key.
 */
export function openWorld(seedPhrase: string): World {
  const worldSeed = hmacSha256(utf8ToBytes(GLOBAL_SALT), utf8ToBytes(seedPhrase));
  return generateWorld(makeMldsaPair(worldSeed), worldSeed);
}

/** Build a World from a signing key and the world seed (mirrors the spec). */
export function generateWorld(signingKey: KeyPair, worldSeed: Uint8Array): World {
  return {
    signingKey,
    walletSalt: hmacSha256(worldSeed, utf8ToBytes("wallets")),
    typeSalt: hmacSha256(worldSeed, utf8ToBytes("types")),
    audienceSalt: hmacSha256(worldSeed, utf8ToBytes("channels")),
  };
}
